src/ConstructionBase.jl

```diff
@@ -95,7 +96,8 @@ end
     Tuple(vals)
 end
 # names are symbols: return namedtuple
-@inline tuple_or_ntuple(::Type{Symbol}, names, vals) = NamedTuple{Tuple(names)}(vals)
+@inline tuple_or_ntuple(::Type{Symbol}, names, vals) = isa(vals, Tuple) ? namedtuple(names, vals...) : NamedTuple{Tuple(names)}(vals)
+@inline namedtuple(names, vals...) = NamedTuple{Tuple(names)}((vals...,)) # this seemingly unnecessary method encourages union splitting.
```

Comment:
It's a bit tricky, it's essentially a transparent performance improvement and allocation-wise it doesn't change much; allocations still happen, e.g. with the `S2` test struct on 1.12 it goes from 64 to 32 bytes allocated, and probably this isn't reliable across Julia versions.

I added the comment in the code in the hope that it won't get changed without reason, because it's hard to catch in tests, but I'm open to suggestions.